### PR TITLE
The long lost cousin of takeExtension has found its way home

### PR DESCRIPTION
**Pull request by SwiftsNamesake**
I compiled and ran the tests. I'm almost sorry it's done. I was hoping to bikeshed at least until Christmas.

### CODE REVIEW

**Comment by ndmitchell** [comment]
Thanks for the pull request. Don't worry, the bike shed is not over!

I think it should work as expected for both ".jpg" and ".tar.gz". Neither seem too hard to add. Can we also get a test that "foo/bar" is not an extension of "hi/foo/bar" as that makes it clear it's not just isSuffixOf.

**Comment by SwiftsNamesake** [comment]
Regarding the first point, are you saying we should add _tests_ for `.jpg` or that `.jpg` currently does not work? Because it _should_, unless I missed something. Or are you saying that we should allow leading dots?

As for ".tar.gz" I mention in the haddock that only the last extension is considered, but you're right that it's a trivial change. I noticed that there were separate functions for one/multiple extensions in other places and `takeExtension` is really `takeFinalExtension` so I'm not entirely sure which behaviour makes the most sense.

I'll go ahead and add tests for "hi/foo/bar" and commit as soon as you've responded to the other points.

File: System/FilePath/Internal.hs [hunk 1]
```diff
@@ -313,6 +313,15 @@ hasExtension :: FilePath -> Bool
 hasExtension = any isExtSeparator . takeFileName
 
 
+-- | Is the given string the final extension of the filename?
+--   The extension should not include the separator.
+--
+-- > "png" `isExtensionOf` "/directory/file.png"     == True
```
Comment (hvr) [inline System/FilePath/Internal.hs]: @ndmitchell this is general question orthogonal to this PR: have you considered using doctest syntax? I.e.

```hs
-- >>> "png" `isExtensionOf` "/directory/file.png"
-- True
```

? I'm just asking, because we're slowly changing the markup in `base` to use that syntax, and I'm looking forward to cabal's upcoming native `cabal new-doctest` support...

Comment (SwiftsNamesake) [inline System/FilePath/Internal.hs]: Oh yeah, I considered it. It's certainly prettier than the == approach, but I noticed that the rest of the module used that so I went along with it.

Comment (hvr) [inline System/FilePath/Internal.hs]: @SwiftsNamesake yeah, it makes totally sense that you followed the existing convention; hence why I directed the question at @ndmitchell :-)

**Comment by SwiftsNamesake** [comment]
@ndmitchell Holding out for your answer, Mitchell.

**Comment by ndmitchell** [comment]
The doctest syntax is more verbose both in source and when rendered. Given the vast quantity of such expressions, that decreases readability. I also dont think it will allow Cabal doc test to work, as my doc tests are somewhat non standard - they test against two modules, apart from when they don't etc.

**Comment by ndmitchell** [comment]
We should add tests for .jpg as the extension because I think it should work, I think users should know it works and I'm not convinced it does. I would expect asking on bar/foo.tar.gz would return True for gz, .gz, tar.gz, .tar.gz

**Comment by SwiftsNamesake** [comment]
@ndmitchell The current implementation  uses `takeExtension`, which gives you the very last extension, including the dot.

@ndmitchell @hvr Would you mind if we took this discussion back to the issue page? I was under the impression that we were in agreement about the 'spec', but it seems now that we're not.

This is the one I had in mind, as did @hvr (his code from the issue page):

```haskell
filter (\fn -> takeExtension fn == ".foo")
filter ((== ".foo") . takeExtension)
```

I suppose we could add two versions (one which is more straight-forward, and another that is more flexible and forgiving).

Once we're in agreement about the desired behaviour, I'll happily implement it, update the tests, and resubmit.

**Comment by ndmitchell** [comment]
Two remarks:

* The commentary on why each test should pass/fail makes the documentation a lot longer, without adding much. I'd just leave the examples and let them speak for themselves. If you think there are anythings worth calling out specifically, move it to the top of the comment. Perhaps saying at the top that that `foo.tar.gz` returns true for all of `gz`, `.gz`, `tar.gz` and `.tar.gz` makes all the corner cases very clear.

**Comment by ndmitchell** [comment]
Second remark:

* You have to generate the tests and then not edit them. You seem to have moved them around and added blank lines before/after. If it isn't exactly as generated, the tests will fail.

**Comment by SwiftsNamesake** [comment]
Sorry about that, didn't read the header comment... Will Travis run `Generate.hs` automatically then?

**Comment by SwiftsNamesake** [comment]
Also, would you mind if I added a How to contribute section in the readme as a separate pull request (unless one of the maintainers want to do it, which frankly would be preferable imho).

**Comment by ndmitchell** [comment]
Travis runs Generate automatically, and if your checked in Generate.hs doesn't perfectly match the one it generates, it will raise an error. Feel free to add a section on how to contribute - I tend to keep a general guide at https://github.com/ndmitchell/neil#contributions, but things like how to generate/test this project are always useful.

**Comment by SwiftsNamesake** [comment]
Apparently you owe me a beer... :p

**Comment by ndmitchell** [comment]
:beers: (can do the same in real life if you're ever passing through my parts of the UK)

**Comment by ndmitchell** [comment]
Thanks for the code, I regenerated the tests, which is a bit subtle. To make it easier in future, I've added some notes at the bottom of the README. Let me know if anything remains unclear.

**Comment by SwiftsNamesake** [comment]
Happy to help. Thanks for the dev notes.
